### PR TITLE
add credential check

### PR DESCRIPTION
**Pull request by runzexia**
In the current version, doCheckCredentialsId does not check whether there is read permission of SCM. This PR mainly completes the function of doCheckCredentialsId.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [ ] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by runzexia** [comment]
@jenkinsci/code-reviewers

**Comment by runzexia** [comment]
cc @MarkEWaite

**Comment by MarkEWaite** [comment]
I don't understand the problem you're trying to solve or the bug you're trying to fix with this pull request.  Please describe:

* use case which is behaving incorrectly
* how you want it to behave
* why this change is the preferred solution for that use case

**Comment by runzexia** [comment]
> I don't understand the problem you're trying to solve or the bug you're trying to fix with this pull request. Please describe:
> 
> * use case which is behaving incorrectly
> * how you want it to behave
> * why this change is the preferred solution for that use case

In the current version, doCheckCredentialsId does not check whether there is read permission of Git repo. This PR mainly completes the function of doCheckCredentialsId.

The behavior I expect is that doCheckCredentialsId can check if jenkins has read access to Git repo.

File: src/test/java/jenkins/plugins/git/GitSCMSourceTest.java [hunk 1]
```diff
@@ -367,6 +365,13 @@ public void gitSCMSourceShouldResolveToolsForMaster() throws Exception {
         assertTrue("Installer should be invoked", inst.isInvoked());
     }
 
+    @Test
+    public void checkCredentialIdFormValidationOk () throws Exception{
+        GitSCMSource.DescriptorImpl scmSource = new GitSCMSource.DescriptorImpl();
+        FormValidation validation = scmSource.doCheckCredentialsId(null,"https://github.com/jenkinsci/git-plugin","");
+        Assert.assertEquals(validation,FormValidation.ok());
+    }
+
```
Comment (MarkEWaite) [inline src/test/java/jenkins/plugins/git/GitSCMSourceTest.java]: Thanks for adding the test.

Please add a test of the negative case as well.  

You're welcome to use 'https://github.com/jenkinsci-cert/git-plugin.git` and `git@github.com:jenkinsci-cert/git-plugin.git` as repositories which exist but are not readable to others.  Alternately, you can refer to a repository that does not exist (since the response is likely the same) as in https://github.com/jenkinsci/no-such-repo.git

Comment (MarkEWaite) [reply src/test/java/jenkins/plugins/git/GitSCMSourceTest.java]: Since the first two of those repositories can be read by me (and hardly anyone else), I'll likely extend the test to use a different repository if the test detects that it is running as me.  You can write the test as yourself, since you don't have access to the jenkinsci-cert repository, and then I can adapt it (if needed) to run in my environment.

**Comment by runzexia** [comment]
add test case

**Comment by runzexia** [comment]
cc @MarkEWaite

File: src/main/java/jenkins/plugins/git/GitSCMSource.java [hunk 2]
```diff
@@ -23,22 +23,23 @@
...
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.EnvVars;
 import hudson.Extension;
 import hudson.RestrictedSince;
 import hudson.Util;
-import hudson.model.Descriptor;
-import hudson.model.Item;
-import hudson.model.ParameterValue;
-import hudson.model.Queue;
+import hudson.model.*;
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitSCMSource.java]: Please don't use wildcard imports.

**Comment by runzexia** [comment]
@MarkEWaite Could you help me to review pr again

**Comment by MarkEWaite** [comment]
@runzexia I added more test cases to cover more of the changes with tests.

* [ ] I need to test the changes interactively.  

I would like to also ask for help from @stephenc to assess if the change from not evaluating credentials in the user interface to evaluating credentials in the user interface is safe enough.

@runzexia, please review the 3 changes I made.  If they are not acceptable, you can either adjust them or use a "force push" to delete them.

**Comment by MarkEWaite** [comment]
@daniel-beck or @Wadeck is there a relevant risk of an amplification attack by adding this check for read permission when a user is defining credentials in the Jenkins job definition page?

**Comment by runzexia** [comment]
@daniel-beck @MarkEWaite  Excuse me, is there any progress now?

**Comment by Wadeck** [comment]
@MarkEWaite sorry no time to review it right now

**Comment by darxriggs** [comment]
@runzexia @MarkEWaite there is a conflict that should be resolved.

**Comment by darxriggs** [comment]
@Wadeck is there a chance that you can have a look at this in the next days?

**Comment by runzexia** [comment]
@darxriggs fix

File: src/main/java/jenkins/plugins/git/GitSCMSource.java [hunk 3]
```diff
@@ -456,35 +467,20 @@ public FormValidation doCheckCredentialsId(@AncestorInPath Item context,
...
 
             remote = Util.fixEmptyAndTrim(remote);
             if (remote == null)
             // not set, can't check
             {
                 return FormValidation.ok();
             }
+            StandardCredentials credentials = value == null ? null :
+                    CredentialsMatchers.firstOrNull(CredentialsProvider
+                                    .lookupCredentials(StandardCredentials.class, context, ACL.SYSTEM,
```
Comment (Wadeck) [inline src/main/java/jenkins/plugins/git/GitSCMSource.java]: ❌ I just looked quickly at the code, no time for deeper investigation right now.

It seems that the removal of 
```
context instanceof Queue.Task	
? Tasks.getAuthenticationOf((Queue.Task) context)	
: ACL.SYSTEM
```
 means that you do not support the `authorize-project` plugin as you do not respect the potential authorization restriction put in place and could be a security vulnerability in such case.

**Comment by darxriggs** [comment]
@runzexia @MarkEWaite there is again a conflict that should be resolved.
